README
mate-dialogs is a fork of gdialog

Zen Understanding -
===================
This is matedialog - a rewrite of gdialog, the MATE port of dialog
which allows you to display dialog boxes from the commandline
and shell scripts. If you understand, things are just as they 
are. If you don't understand, things are just as they are.

This software is licensed under the LGPL.

Zen Requirements -
==================
scrollkeeper and intltool.
gtk+-2.0


NEWS
MateDialogs 1.8.0
  * Added GTK3 support
  * Use yelp-tools instead of mate-doc-utils

MateDialogs 1.6.0
  * Use libnotify instead of libmatenofity




src/util.c
/*
 * util.c
 *
 * Copyright (C) 2002 Sun Microsystems, Inc.
 *           (C) 1999, 2000 Red Hat Inc.
 *           (C) 1998 James Henstridge
 *           (C) 1995-2002 Free Software Foundation
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 *
 * Authors: Glynn Foster <glynn.foster@sun.com>
 *          Havoc Pennington <hp@redhat.com>
 *          James Henstridge <james@daa.com.au>
 *          Tom Tromey <tromey@redhat.com>
 */

#include "config.h"

#include <stdio.h>
#include <locale.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include "config.h"
#include "util.h"
#include "matedialog.h"

#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#define MATEDIALOG_OK_DEFAULT	0
#define MATEDIALOG_CANCEL_DEFAULT	1
#define MATEDIALOG_ESC_DEFAULT	1
#define MATEDIALOG_ERROR_DEFAULT	-1
#define MATEDIALOG_EXTRA_DEFAULT	127

GtkBuilder*
matedialog_util_load_ui_file (const gchar *root_widget, ...)
{
  va_list args;
  gchar *arg = NULL;
  GPtrArray *ptrarray;
  GtkBuilder *builder = gtk_builder_new ();
  GError *error = NULL;
  gchar  **objects;
  guint result = 0;

  gtk_builder_set_translation_domain (builder, GETTEXT_PACKAGE);

  /* We have at least the root_widget and a NULL */
  ptrarray = g_ptr_array_sized_new (2);

  g_ptr_array_add (ptrarray, g_strdup (root_widget));

  va_start (args, root_widget);

  arg = va_arg (args, gchar*);

  while (arg) {
  	g_ptr_array_add (ptrarray, g_strdup (arg));
	arg = va_arg (args, gchar*);
  }
  va_end (args);

  /* Enforce terminating NULL */
  g_ptr_array_add (ptrarray, NULL);
  objects = (gchar**) g_ptr_array_free (ptrarray, FALSE);

  if (g_file_test (MATEDIALOG_UI_FILE_RELATIVEPATH, G_FILE_TEST_EXISTS)) {
    /* Try current dir, for debugging */
    result = gtk_builder_add_objects_from_file (builder,
    						MATEDIALOG_UI_FILE_RELATIVEPATH,
						objects, NULL);
  }

  if (result == 0)
    result = gtk_builder_add_objects_from_file (builder,
    						MATEDIALOG_UI_FILE_FULLPATH,
						objects, &error);

  g_strfreev (objects);

  if (result == 0) {
    g_warning ("Could not load ui file %s: %s", MATEDIALOG_UI_FILE_FULLPATH,
    						error->message);
    g_error_free (error);
    g_object_unref (builder);
    return NULL;
  }

  return builder;
}
gchar*
matedialog_util_strip_newline (gchar *string)
{
    gsize len;
    
    g_return_val_if_fail (string != NULL, NULL);
                                                                                                                                                                             
    len = strlen (string);
    while (len--) 
    {
      if (string[len] == '\n')
        string[len] = '\0';
      else
        break;
    }
            
    return string;
}

gboolean
matedialog_util_fill_file_buffer (GtkTextBuffer *buffer, const gchar *filename) 
{
  GtkTextIter iter, end;
  FILE *f;
  gchar buf[2048];
  gint remaining = 0;

  if (filename == NULL)
    return FALSE;

  f = fopen (filename, "r");

  if (f == NULL) {
    g_warning ("Cannot open file '%s': %s", filename, g_strerror (errno));
    return FALSE;
  }

  gtk_text_buffer_get_iter_at_offset (buffer, &iter, 0);

  while (!feof (f)) {
    gint count;
    const char *leftover;
    int to_read = 2047  - remaining;

    count = fread (buf + remaining, 1, to_read, f);
    buf[count + remaining] = '\0';

    g_utf8_validate (buf, count + remaining, &leftover);

    g_assert (g_utf8_validate (buf, leftover - buf, NULL));
    gtk_text_buffer_insert (buffer, &iter, buf, leftover - buf);

    remaining = (buf + remaining + count) - leftover;
    g_memmove (buf, leftover, remaining);

    if (remaining > 6 || count < to_read)
      break;
  }

  if (remaining) {
    g_warning ("Invalid UTF-8 data encountered reading file '%s'", filename);
    return FALSE;
  }

  /* We had a newline in the buffer to begin with. (The buffer always contains
   * a newline, so we delete to the end of the buffer to clean up.
   */
  
  gtk_text_buffer_get_end_iter (buffer, &end);
  gtk_text_buffer_delete (buffer, &iter, &end);

  gtk_text_buffer_set_modified (buffer, FALSE);

  return TRUE;
}

const gchar *
matedialog_util_stock_from_filename (const gchar *filename)
{
  if (!filename || !filename[0])
    return GTK_STOCK_DIALOG_WARNING; /* default */

  if (!g_ascii_strcasecmp (filename, "warning"))
    return GTK_STOCK_DIALOG_WARNING;
  if (!g_ascii_strcasecmp (filename, "info"))
    return GTK_STOCK_DIALOG_INFO;
  if (!g_ascii_strcasecmp (filename, "question"))
    return GTK_STOCK_DIALOG_QUESTION;
  if (!g_ascii_strcasecmp (filename, "error"))
    return GTK_STOCK_DIALOG_ERROR;
  return NULL;
}

GdkPixbuf *
matedialog_util_pixbuf_new_from_file (GtkWidget *widget, const gchar *filename)
{
  const gchar *stock;

  stock = matedialog_util_stock_from_filename (filename);
  if (stock)
    return gtk_widget_render_icon (widget, stock, GTK_ICON_SIZE_BUTTON, NULL);

 return gdk_pixbuf_new_from_file (filename, NULL);
}

void
matedialog_util_set_window_icon (GtkWidget *widget, const gchar *filename, const gchar *default_file)
{
  GdkPixbuf *pixbuf;

  if (filename != NULL)
    pixbuf = matedialog_util_pixbuf_new_from_file (widget, (gchar *) filename);
  else
    pixbuf = gdk_pixbuf_new_from_file (default_file, NULL);  

  if (pixbuf != NULL) {
    gtk_window_set_icon (GTK_WINDOW (widget), pixbuf);
    g_object_unref (pixbuf);
  }
}

void 
matedialog_util_set_window_icon_from_stock (GtkWidget *widget, const gchar *filename, const gchar *default_stock_id)
{
  GdkPixbuf *pixbuf;

  if (filename != NULL) {
    pixbuf = matedialog_util_pixbuf_new_from_file (widget, (gchar *) filename);
  }
  else {
    pixbuf = gtk_widget_render_icon (widget, default_stock_id, GTK_ICON_SIZE_BUTTON, NULL);
  }

  if (pixbuf != NULL) {
    gtk_window_set_icon (GTK_WINDOW (widget), pixbuf);
    g_object_unref (pixbuf);
  }
}

void
matedialog_util_show_help (GError **error)
{
  gchar *tmp;
  tmp = g_find_program_in_path ("yelp");

  if (tmp) {
    g_free (tmp);
    g_spawn_command_line_async ("yelp help:matedialog", error);
  }
}

gint 
matedialog_util_return_exit_code ( MateDialogExitCode value ) 
{

  const gchar *env_var = NULL;
  gint retval;

  switch (value) {
  
  case MATEDIALOG_OK:
    env_var = g_getenv("MATEDIALOG_OK");
    if (! env_var) 
          env_var = g_getenv("DIALOG_OK");
    if (! env_var) 
          retval = MATEDIALOG_OK_DEFAULT;
    break;
   
  case MATEDIALOG_CANCEL:
    env_var = g_getenv("MATEDIALOG_CANCEL");
    if (! env_var) 
          env_var = g_getenv("DIALOG_CANCEL");
    if (! env_var) 
          retval = MATEDIALOG_CANCEL_DEFAULT;
    break;
    
  case MATEDIALOG_ESC:
    env_var = g_getenv("MATEDIALOG_ESC");
    if (! env_var) 
          env_var = g_getenv("DIALOG_ESC");
    if (! env_var) 
          retval = MATEDIALOG_ESC_DEFAULT;
    break;
    
  case MATEDIALOG_EXTRA:
    env_var = g_getenv("MATEDIALOG_EXTRA");
    if (! env_var) 
          env_var = g_getenv("DIALOG_EXTRA");
    if (! env_var) 
          retval = MATEDIALOG_EXTRA_DEFAULT;
    break;
    
  case MATEDIALOG_ERROR:
    env_var = g_getenv("MATEDIALOG_ERROR");
    if (! env_var) 
          env_var = g_getenv("DIALOG_ERROR");
    if (! env_var) 
          retval = MATEDIALOG_ERROR_DEFAULT;
    break;
    
  default:
    retval = 1;
  }
  
  if (env_var) 
      retval = atoi (env_var);
  return retval; 
}


#ifdef GDK_WINDOWING_X11

static Window
transient_get_xterm (void)
{
  const char *wid_str = g_getenv ("WINDOWID");
  if (wid_str) {
    char *wid_str_end;
    int ret;
    Window wid = strtoul (wid_str, &wid_str_end, 10);
    if (*wid_str != '\0' && *wid_str_end == '\0' && wid != 0) {
      XWindowAttributes attrs;
      gdk_error_trap_push ();
      ret = XGetWindowAttributes (GDK_DISPLAY_XDISPLAY(gdk_display_get_default()), wid, &attrs);
      gdk_flush();
      if (gdk_error_trap_pop () != 0 || ret == 0) {
        return None;
      }
      return wid;
    }
  }
  return None;
}

static void
transient_x_free (void *ptr)
{
  if (ptr)
    XFree (ptr);
}

static gboolean
transient_is_toplevel (Window wid)
{
  XTextProperty prop;
  Display *dpy = GDK_DISPLAY_XDISPLAY(gdk_display_get_default());
  if (!XGetWMName (dpy, wid, &prop))
    return FALSE;
  transient_x_free (prop.value);
  return !!prop.value;
}

/*
 * MATE Terminal doesn't give us its toplevel window, but the WM needs a
 * toplevel XID for proper stacking.  Other terminals work fine without this
 * magic.  We can't use GDK here since "xterm" is a foreign window.
 */

static Window
transient_get_xterm_toplevel (void)
{
  Window xterm = transient_get_xterm ();
  Display *dpy = GDK_DISPLAY_XDISPLAY(gdk_display_get_default());
  while (xterm != None && !transient_is_toplevel (xterm))
  {
    Window root, parent, *children;
    unsigned nchildren;
    XQueryTree (dpy, xterm,
                &root, &parent,
                &children, &nchildren);
    transient_x_free (children);
    if (parent == root)
      xterm = None;
    else
      xterm = parent;
  }
  return xterm;
}

static void
matedialog_util_make_transient (GdkWindow *window)
{
  Window xterm = transient_get_xterm_toplevel ();
  if (xterm != None) {
    #if GTK_CHECK_VERSION(2,24,0)
      GdkWindow *gdkxterm = gdk_x11_window_foreign_new_for_display (
                          gdk_display_get_default (),
                          xterm);
    #else
      GdkWindow *gdkxterm = gdk_window_foreign_new (xterm);
    #endif
    if (gdkxterm) {
      gdk_window_set_transient_for (window, gdkxterm);
      g_object_unref (G_OBJECT (gdkxterm));
    }
  }
}

#endif /* GDK_WINDOWING_X11 */

void
matedialog_util_show_dialog (GtkWidget *dialog)
{
  gtk_widget_realize (dialog);
#ifdef GDK_WINDOWING_X11
  g_assert (gtk_widget_get_window(dialog));
  matedialog_util_make_transient (gtk_widget_get_window(dialog));
#endif
  gtk_widget_show (dialog);
}

gboolean 
matedialog_util_timeout_handle (void)
{
  gtk_main_quit();
  exit(MATEDIALOG_TIMEOUT);
  return FALSE;
}


src/entry.c
/*
 * entry.c
 *
 * Copyright (C) 2002 Sun Microsystems, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 *
 * Authors: Glynn Foster <glynn.foster@sun.com>
 */

#include "config.h"

#include "matedialog.h"
#include "util.h"

static void matedialog_entry_dialog_response (GtkWidget *widget, int response, gpointer data);

static GtkWidget *entry;
static gint n_entries = 0;

static void
matedialog_entry_fill_entries (GSList **entries, const gchar **args)
{
  gint i = 0;

  while (args[i] != NULL) {
    *entries = g_slist_append (*entries, (gchar *) args[i]);
    i++;
  }
}

static void
matedialog_entry_combo_activate_default (GtkEntry *entry, gpointer window)
{
  gtk_window_activate_default (GTK_WINDOW (window));
}

void 
matedialog_entry (MateDialogData *data, MateDialogEntryData *entry_data)
{
  GtkBuilder *builder = NULL;
  GtkWidget *dialog;
  GObject *text;
  GSList *entries = NULL; 
  GSList *tmp;
  GObject *vbox;
  
  builder = matedialog_util_load_ui_file ("matedialog_entry_dialog", NULL);

  if (builder == NULL) {
    data->exit_code = matedialog_util_return_exit_code (MATEDIALOG_ERROR);
    return;
  }
	
  gtk_builder_connect_signals (builder, NULL);
	
  dialog = GTK_WIDGET (gtk_builder_get_object (builder, "matedialog_entry_dialog"));
	
  g_signal_connect (G_OBJECT (dialog), "response",
                    G_CALLBACK (matedialog_entry_dialog_response), data);

  if (data->dialog_title)
    gtk_window_set_title (GTK_WINDOW (dialog), data->dialog_title);
	
  matedialog_util_set_window_icon (dialog, data->window_icon, MATEDIALOG_IMAGE_FULLPATH ("matedialog-entry.png"));

  if (data->width > -1 || data->height > -1)
    gtk_window_set_default_size (GTK_WINDOW (dialog), data->width, data->height);

  text = gtk_builder_get_object (builder, "matedialog_entry_text");

  if (entry_data->dialog_text)
    gtk_label_set_text_with_mnemonic (GTK_LABEL (text), g_strcompress (entry_data->dialog_text));
  
  vbox = gtk_builder_get_object (builder, "vbox4");
  
  matedialog_entry_fill_entries(&entries, entry_data->data);
  
  n_entries = g_slist_length (entries);

  if (n_entries > 1) {
    #if GTK_CHECK_VERSION(2,24,0)
      entry = gtk_combo_box_text_new_with_entry ();
    #else
      entry = gtk_combo_box_entry_new_text ();
    #endif

    for (tmp = entries; tmp; tmp = tmp->next) {
      #if GTK_CHECK_VERSION(2,24,0)
        gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (entry), tmp->data);
      #else
        gtk_combo_box_append_text (GTK_COMBO_BOX (entry), tmp->data);
      #endif
    }

    if (entry_data->entry_text) {
      #if GTK_CHECK_VERSION(2,24,0)
        gtk_combo_box_text_prepend_text (GTK_COMBO_BOX_TEXT (entry), entry_data->entry_text);
      #else
        gtk_combo_box_prepend_text (GTK_COMBO_BOX (entry), entry_data->entry_text);
      #endif
      gtk_combo_box_set_active (GTK_COMBO_BOX (entry), 0);
    }

    g_signal_connect (gtk_bin_get_child (GTK_BIN (entry)), "activate",
                      G_CALLBACK (matedialog_entry_combo_activate_default),
                      GTK_WINDOW (dialog));
  } else {
    entry = gtk_entry_new();

    gtk_entry_set_activates_default (GTK_ENTRY (entry), TRUE);
    
    if (entry_data->entry_text)
      gtk_entry_set_text (GTK_ENTRY (entry), entry_data->entry_text);

    if (entry_data->hide_text)
      g_object_set (G_OBJECT (entry), "visibility", FALSE, NULL);
  }

  gtk_widget_show (entry);

  gtk_box_pack_end (GTK_BOX (vbox), entry, FALSE, FALSE, 0);

  gtk_label_set_mnemonic_widget (GTK_LABEL (text), entry);

  g_object_unref (builder);

  matedialog_util_show_dialog (dialog);

  if(data->timeout_delay > 0) {
    g_timeout_add_seconds (data->timeout_delay, (GSourceFunc) matedialog_util_timeout_handle, NULL);
  }

  gtk_main ();
}

static void
matedialog_entry_dialog_response (GtkWidget *widget, int response, gpointer data)
{
  MateDialogData *zen_data = data;
  const gchar *text;

  switch (response) {
    case GTK_RESPONSE_OK:
      zen_data->exit_code = matedialog_util_return_exit_code (MATEDIALOG_OK);
      if (n_entries > 1) {
        #if GTK_CHECK_VERSION(2,24,0)
          text = gtk_combo_box_text_get_active_text (GTK_COMBO_BOX_TEXT (entry));
        #else
          text = gtk_combo_box_get_active_text (GTK_COMBO_BOX (entry));
        #endif
      }
      else {
        text = gtk_entry_get_text (GTK_ENTRY (entry));      
      }

      if (text != NULL)
        g_print ("%s\n", text);

      break;

    case GTK_RESPONSE_CANCEL:
      zen_data->exit_code = matedialog_util_return_exit_code (MATEDIALOG_CANCEL);
      break;

    default:
      /* Esc dialog */
      zen_data->exit_code = matedialog_util_return_exit_code (MATEDIALOG_ESC);
      break;
  }
  gtk_main_quit ();
}


src/option.h
/*
 * option.h
 *
 * Copyright (C) 2002 Sun Microsystems, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 *
 * Authors: Glynn Foster <glynn.foster@sun.com>
 *          Lucas Rocha <lucasr@im.ufba.br>
 */

#ifndef OPTION_H
#define OPTION_H

#include "matedialog.h"
#include <glib.h>
#include <langinfo.h>
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif

typedef enum {
  MODE_CALENDAR,
  MODE_ENTRY,
  MODE_ERROR,
  MODE_FILE,
  MODE_LIST,
  MODE_PROGRESS,
  MODE_QUESTION,
  MODE_TEXTINFO,
  MODE_WARNING,
  MODE_SCALE,
  MODE_INFO,
  MODE_NOTIFICATION,
  MODE_COLOR,
  MODE_PASSWORD,
  MODE_ABOUT,
  MODE_VERSION,
  MODE_LAST
} MateDialogDialogMode;

typedef enum {
  ERROR_SYNTAX,
  ERROR_SUPPORT,
  ERROR_DIALOG,
  ERROR_LAST
} MateDialogError;

typedef struct {
  MateDialogDialogMode mode;
  MateDialogData      *data;

  MateDialogCalendarData     *calendar_data;
  MateDialogMsgData          *msg_data;
  MateDialogScaleData        *scale_data;
  MateDialogFileData         *file_data;
  MateDialogEntryData        *entry_data;
  MateDialogProgressData     *progress_data;
  MateDialogTextData         *text_data;
  MateDialogTreeData         *tree_data;
  MateDialogNotificationData *notification_data;
  MateDialogColorData        *color_data;
  MateDialogPasswordData     *password_data;
} MateDialogParsingOptions;

void			matedialog_option_error (gchar	*string,
					     MateDialogError error);

MateDialogParsingOptions * 	matedialog_option_parse (gint	 argc,
					     gchar     **argv); 

void 			matedialog_option_free (void);

#endif /* OPTION_H */


src/about.c
/*
 * about.c
 *
 * Copyright (C) 2002 Sun Microsystems, Inc.
 * Copyright (C) 2001 CodeFactory AB
 * Copyright (C) 2001, 2002 Anders Carlsson
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 *
 * Authors: Glynn Foster <glynn.foster@sun.com>
 *          Anders Carlsson <andersca@gnu.org>
 */

#include "config.h"
#include "matedialog.h"
#include "util.h"
#include <string.h>
#include <gdk/gdkkeysyms.h>

#define GTK_RESPONSE_CREDITS 0
#define MATEDIALOG_HELP_PATH MATEDIALOG_DATADIR "/help/"
#define MATEDIALOG_CLOTHES_PATH MATEDIALOG_DATADIR "/clothes/"

#define MATEDIALOG_CANVAS_X 400.0
#define MATEDIALOG_CANVAS_Y 280.0

static GtkWidget *dialog;

static void matedialog_about_display_help (GtkWidget *widget, gpointer data);
static void matedialog_about_dialog_response (GtkWidget *widget, int response, gpointer data);

/* Sync with the people in the THANKS file */
static const gchar *const authors[] = {
  "Glynn Foster <glynn foster sun com>",
  "Lucas Rocha <lucasr mate org>",
  "Mike Newman <mikegtn mate org>",
  NULL
};

static const char *documenters[] = {
  "Glynn Foster <glynn.foster@sun.com>",
  "Lucas Rocha <lucasr@gnome.org>",
  "Java Desktop System Documentation Team",
  "MATE Documentation Project",
  NULL
};

static gchar *translators;

static const char *license[] = {
	N_("This program is free software; you can redistribute it and/or modify "
	   "it under the terms of the GNU Lesser General Public License as published by "
	   "the Free Software Foundation; either version 2 of the License, or "
	   "(at your option) any later version.\n"),
	N_("This program is distributed in the hope that it will be useful, "
	   "but WITHOUT ANY WARRANTY; without even the implied warranty of "
	   "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the "
	   "GNU Lesser General Public License for more details.\n"),
	N_("You should have received a copy of the GNU Lesser General Public License "
	   "along with this program; if not, write to the Free Software "
	   "Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.")
};

#if 0
static gint
matedialog_move_clothes_event (MateCanvasItem *item,
                           GdkEvent *event,
                           gpointer data)
{
  static double x, y;
  double new_x, new_y;
  static int dragging;
  double item_x, item_y;

  /* set item_[xy] to the event x,y position in the parent's
   * item-relative coordinates
   */

  item_x = event->button.x;
  item_y = event->button.y;
  mate_canvas_item_w2i (item->parent, &item_x, &item_y);

  switch (event->type) {
    case GDK_BUTTON_PRESS:
      x = item_x;
      y = item_y;
      mate_canvas_item_ungrab (item, event->button.time);
      mate_canvas_item_raise_to_top (item);
      dragging = TRUE;
      break;

    case GDK_MOTION_NOTIFY:
      if (dragging && (event->motion.state & GDK_BUTTON1_MASK)) {
        new_x = item_x;
        new_y = item_y;

        mate_canvas_item_move (item, new_x - x, new_y - y);
        x = new_x;
        y = new_y;
      }
      break;

    case GDK_BUTTON_RELEASE:
      mate_canvas_item_ungrab (item, event->button.time);
      dragging = FALSE;
      break;

    default:
      break;
  }

  return FALSE;
}

typedef struct
{
  const gchar *filename;
  gdouble x, y;
} MonkClothes;

static const MonkClothes monk_clothes[] = {
  {"mate-tshirt.png", 30.0, 20.0},
  {"sunglasses.png", MATEDIALOG_CANVAS_X - 100.0 , MATEDIALOG_CANVAS_Y - 150.0 },
  {"surfboard.png", 30.0, MATEDIALOG_CANVAS_Y - 200.0},
  {"hawaii-shirt.png", MATEDIALOG_CANVAS_X - 50.0, 20.0}
};

static void
matedialog_create_clothes (GtkWidget *canvas_board)
{
  GdkPixbuf *pixbuf;
  MateCanvasItem *canvas_item;
  gchar *pixbuf_path;
  size_t i;

  for (i = 0; i < G_N_ELEMENTS (monk_clothes); i++) {
    pixbuf_path = g_strconcat (MATEDIALOG_CLOTHES_PATH, monk_clothes[i].filename, NULL);
    pixbuf = gdk_pixbuf_new_from_file (pixbuf_path, NULL);

    canvas_item = mate_canvas_item_new (MATE_CANVAS_GROUP (MATE_CANVAS (canvas_board)->root),
                                         MATE_TYPE_CANVAS_PIXBUF,
                                         "x", monk_clothes[i].x,
                                         "y", monk_clothes[i].y,
                                         "pixbuf", pixbuf,
                                         "anchor", GTK_ANCHOR_NW,
                                          NULL);
    g_signal_connect (G_OBJECT (canvas_item), "event",
                      G_CALLBACK (matedialog_move_clothes_event), NULL);
  }
}

static GtkWidget *
matedialog_create_monk (void)
{
  GtkWidget *canvas_board;
  MateCanvasItem *canvas_item;
  GdkPixbuf *pixbuf;
  GdkColor color = { 0, 0xffff, 0xffff, 0xffff };

  canvas_board = mate_canvas_new ();

  mate_canvas_set_scroll_region (MATE_CANVAS (canvas_board), 0, 0,
                                  MATEDIALOG_CANVAS_X, MATEDIALOG_CANVAS_Y);

  gtk_widget_set_size_request (canvas_board, MATEDIALOG_CANVAS_X, MATEDIALOG_CANVAS_Y);

  gdk_colormap_alloc_color (gtk_widget_get_colormap (GTK_WIDGET (canvas_board)),
                            &color, FALSE, TRUE);

  gtk_widget_modify_bg (GTK_WIDGET (canvas_board), GTK_STATE_NORMAL, &color);

  pixbuf = gdk_pixbuf_new_from_file (MATEDIALOG_CLOTHES_PATH "monk.png", NULL);

  canvas_item = mate_canvas_item_new (MATE_CANVAS_GROUP (MATE_CANVAS (canvas_board)->root),
                                       MATE_TYPE_CANVAS_PIXBUF,
                                       "x", (MATEDIALOG_CANVAS_X / 2.0)/2.0 + 10.0,
                                       "y", (MATEDIALOG_CANVAS_Y / 2.0)/2.0 - 50.0,
                                       "pixbuf", pixbuf,
                                       "anchor", GTK_ANCHOR_NW,
                                       NULL);

  matedialog_create_clothes (canvas_board);

  return canvas_board;
}

static GtkWidget *
matedialog_create_boutique (void)
{
  GtkWidget *window;
  GtkWidget *canvas;

  window = gtk_window_new (GTK_WINDOW_TOPLEVEL);

  matedialog_util_set_window_icon (window, NULL, MATEDIALOG_IMAGE_FULLPATH ("matedialog.png"));
  canvas = matedialog_create_monk ();
  gtk_container_add (GTK_CONTAINER (window), canvas);

  return window;
}

static gboolean
matedialog_zen_wisdom (GtkDialog *dialog, GdkEventKey *event, gpointer user_data)
{
  static gint string_count;

  if (string_count >= 3)
    return FALSE;

  switch (event->keyval) {
    case GDK_N:
    case GDK_n:
      if (string_count == 2) {
        GtkWidget *window;
        window = matedialog_create_boutique ();
        gtk_widget_show_all (window);
        string_count++;
      } else {
        string_count = 0;
      }
      break;
    case GDK_Z:
    case GDK_z:
      if (string_count == 0)
        string_count++;
      else
        string_count = 0;
      break;
    case GDK_E:
    case GDK_e:
      if (string_count == 1)
        string_count++;
      else
        string_count = 0;
      break;
    default:
      string_count = 0;
  }

  return FALSE;
}
#endif

void
matedialog_about (MateDialogData *data)
{
  GdkPixbuf *logo;
  GtkWidget *help_button;
  char *license_trans;


  translators = _("translator-credits");
  logo = gdk_pixbuf_new_from_file (MATEDIALOG_IMAGE_FULLPATH ("matedialog.png"), NULL);

  license_trans = g_strconcat (_(license[0]), "\n", _(license[1]), "\n",
                               _(license[2]), "\n", NULL);

  dialog = gtk_about_dialog_new ();

  g_object_set (G_OBJECT (dialog),
                "name", "MateDialog",
                "version", VERSION,
                "copyright", "Copyright \xc2\xa9 2003 Sun Microsystems",
                "comments", _("Display dialog boxes from shell scripts"),
		"authors", authors,
                "documenters", documenters,
                "translator-credits", translators,
                "website", "http://www.mate-desktop.org/",
                "logo", logo,
                "wrap-license", TRUE,
                "license", license_trans,
		NULL);

  g_free (license_trans);

  matedialog_util_set_window_icon (dialog, NULL, MATEDIALOG_IMAGE_FULLPATH ("matedialog.png"));

  help_button = gtk_button_new_from_stock (GTK_STOCK_HELP);

  g_signal_connect (G_OBJECT (help_button), "clicked",
                    G_CALLBACK (matedialog_about_display_help), data);

  gtk_widget_show (help_button);

  gtk_box_pack_end (GTK_BOX (gtk_dialog_get_action_area (GTK_DIALOG (dialog))),
                    help_button, FALSE, TRUE, 0);
  gtk_button_box_set_child_secondary (GTK_BUTTON_BOX (gtk_dialog_get_action_area (GTK_DIALOG (dialog))),
                                      help_button, TRUE);

  g_signal_connect (G_OBJECT (dialog), "response",
                    G_CALLBACK (matedialog_about_dialog_response), data);

#if 0
  g_signal_connect (G_OBJECT (dialog), "key_press_event",
                    G_CALLBACK (matedialog_zen_wisdom), NULL);
#endif

  matedialog_util_show_dialog (dialog);
  gtk_main ();
}

static void
matedialog_about_dialog_response (GtkWidget *widget, int response, gpointer data)
{
  MateDialogData *zen_data = data;

  switch (response) {
    case GTK_RESPONSE_CLOSE:
      zen_data->exit_code = matedialog_util_return_exit_code (MATEDIALOG_OK);
      break;

    default:
      /* Esc dialog */
      zen_data->exit_code = matedialog_util_return_exit_code (MATEDIALOG_ESC);
      break;
  }

  gtk_main_quit ();
}

static void
matedialog_about_display_help (GtkWidget *widget, gpointer data)
{
  matedialog_util_show_help (NULL);
}
